sway-core/src/semantic_analysis/cei_pattern_analysis.rs

```diff
@@ -0,0 +1,258 @@
+// CEI stands for "Checks, Effects, Interactions".  We check that no storage writes
+// (effects) occur after calling external contracts (interaction) and issue warnings
+// if it's the case.
+// See this [blog post](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html)
+// for more detail.
+
+use crate::{
+    declaration_engine::DeclarationId,
+    language::ty::{self, TyFunctionDeclaration},
+};
+use std::collections::HashSet;
+use sway_error::warning::{CompileWarning, Warning};
+use sway_types::Span;
+
+#[derive(PartialEq, Eq, Hash, Clone)]
+enum Effect {
+    Interaction,  // interaction with external contracts
+    StorageWrite, // storage modification
+}
+
+enum CEIAnalysisState {
+    LookingForInteraction,
+    LookingForStorageWrite,
+}
+
+// TODO: process libraries, scripts and predicates
+pub(crate) fn program(prog: &ty::TyProgram) -> Vec<CompileWarning> {
```

Comment:
Should `program` have a more descriptive name?

Reply:
I suppose the same applies to many functions in the module. I see that you're naming them based on what program entity they run on, but something a bit more descriptive would go a long way here.